### DISCUSSION TITLE
docs: pessimistic lock and auto commit

### DISCUSSION THREAD

**Discussion by axotion** [mikro-orm/mikro-orm#4096] [General]
In docs (https://mikro-orm.io/docs/transactions#pessimistic-locking) we can see statment

>However, for Pessimistic Locking to work you have to disable the Auto-Commit Mode of your Database and start a transaction around your pessimistic lock use-case using the "Approach 2: Explicit Transaction Demarcation" described above. MikroORM will throw an Exception if you attempt to acquire a pessimistic lock and no transaction is running.

But if you are opening transactions manually and use eg. for update

BEGIN;
SELECT balance FROM wallet WHERE id = 1 AND balance >= 100 FOR UPDATE;
UPDATE wallet SET balance = balance - 100 WHERE id = 1;
COMMIT;

Then it should be fine despite enabled auto-commit on the database, right? MikroORM runs on knex.js, which allows something like that. Or maybe I'm wrong and auto-commit will still be a problem even with manual transactions?

**Comment by B4nan** [comment]
I am not sure what are you asking here. You showed only SQL commands but ask about how the ORM behaves - does this mean you are creating all the commands manually or via knex? If so, the ORM docs are irrelevant anyway.

That sentence only means that you can use pessimistic locking only inside transactions, and the ORM guards you from not doing that, it literally throws. If you don't see any errors, you are most probably fine. And if you see a `begin` statement, I don't think auto-commit can have any effect - as it is there to create transactions (on db level) when there are none - you literally disable that by opening a transaction.

**Comment by axotion** [comment]
Sorry, I will try to be more specific

I was not sure about this requirement
>However, for Pessimistic Locking to work you have to disable the Auto-Commit Mode of your Database and start a transaction around your pessimistic lock use-case using the "Approach 2: Explicit Transaction Demarcation"

Where you have to disable Auto-Commit Mode on your Database in order to use MikroORM pessimistic locks. Do I really need to disable auto-commit mode on database in order to use locks with MikroORM? Because I once read that

>Locking of rows for update using SELECT FOR UPDATE only applies when autocommit is disabled (either by beginning transaction with START TRANSACTION or by setting autocommit to 0. If autocommit is enabled, the rows matching the specification are not locked.

And with a note from https://dev.mysql.com/doc/refman/8.0/en/innodb-locking-reads.html
>Locking reads are only possible when autocommit is disabled (either by beginning transaction with [START TRANSACTION](https://dev.mysql.com/doc/refman/8.0/en/commit.html) or by setting [autocommit](https://dev.mysql.com/doc/refman/8.0/en/server-system-variables.html#sysvar_autocommit) to 0.

And now I'm confused, because MikroORM offers transactions on ORM level, so they should be fine with enabled auto-commit on database, right? Or transactions on ORM level != database transaction? 

I know that I'm reffering to INNODB from MySQL and MikroORM has support for other engines, but maybe because of this, MikroORM has only transactions on code level, not on database level?

**Reply by B4nan** [reply]
> when autocommit is disabled (**either by beginning transaction with START TRANSACTION** or by setting autocommit to 0. If autocommit is enabled, the rows matching the specification are not locked.

It gets disabled by starting a transaction, as your quote says. 

I guess the docs are confusing, I based this on doctrine docs as the behavior is the same. I'll remove the note about auto commit, I believe it's not something you should really care about, the information in doctrine docs might be based on some old versions or different drivers.

Re-reading the doctrine docs now, it feels like this information might be actually PHP specific.

**Reply by axotion** [reply]
I believe they also have some kind of mistake in their docs, because

https://www.doctrine-project.org/projects/doctrine-orm/en/2.14/reference/transactions-and-concurrency.html#pessimistic-locking

They have their version of what you wrote in MikroORM docs,

>However for Pessimistic Locking to work you have to disable the Auto-Commit Mode of your Database and start a transaction around your pessimistic lock use-case using the Approach 2: Explicit Transaction Demarcation described above.

**BUT** in Approach 2: Explicit Transaction Demarcation we can see something weird
https://www.doctrine-project.org/projects/doctrine-orm/en/2.14/reference/transactions-and-concurrency.html#approach-2-explicitly

```$em->getConnection()->beginTransaction(); // suspend auto-commit```

Which is contradictory to what they say about autocommits
 >you have to disable the Auto-Commit Mode of your Database and start a transaction
